lib/onelogin/ruby-saml/response.rb

```diff
@@ -112,78 +151,326 @@ def status_message
-          node ||= xpath_first_from_signed_assertion('/a:Issuer')
-          node.nil? ? nil : node.text
+      # Gets the Issuers (from Response and Assertion).
+      # (returns the first node that matches the supplied xpath from the Response and from the Assertion)
+      # @return [Array] Array with the Issuers (REXML::Element)
+      #
+      def issuers
+        @issuers ||= begin
+          issuers = []
+          nodes = REXML::XPath.match(document, "/p:Response/a:Issuer | /p:Response/a:Assertion/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
+          nodes.each do |node|
+            issuers << node.text if not node.text.nil?
+          end
+          issuers.uniq
+        end
+      end
+
+      # @return [String|nil] The InResponseTo attribute from the SAML Response.
+      #
+      def in_response_to
+        @in_response_to ||= begin
+          node = REXML::XPath.first(document, "/p:Response", { "p" => PROTOCOL })
+          node.nil? ? nil : node.attributes['InResponseTo']
         end
       end
 
+      # @return [String|nil] Destination attribute from the SAML Response
+      #
+      def destination
+        @destination ||= begin
+          node = REXML::XPath.first(document, "/p:Response", { "p" => PROTOCOL })
+          node.nil? ? nil : node.attributes['Destination']
+        end
+      end
+
+      # @return [Array] The Audience elements from the Contitions of the SAML Response.
+      #
+      def audiences
+        @audiences ||= begin
+          audiences = []
+          nodes = xpath_from_signed_assertion('/a:Conditions/a:AudienceRestriction/a:Audience')
+          nodes.each do |node|
+            if node && node.text
+              audiences << node.text
+            end
+          end
+          audiences
+        end
+     end
+
       private
 
-      def validate(soft = true)
-        valid_saml?(document, soft)      &&
+      # Gets the expected current_url
+      # (Right now we read this url from the Assertion Consumer Service of the Settings)
+      # TODO: Calculate the real current_url and use it.
+      # @return [String] The current url
+      #
+      def current_url
+        @current_url ||= begin
+          settings.assertion_consumer_service_url
+        end
+      end
+
+      # Validates the SAML Response (calls several validation methods)
+      # If fails, the attribute errors will contains the reason for the invalidation.
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @param request_id [String|nil] request_id The ID of the AuthNRequest sent by this SP to the IdP (if was sent any)
+      # @return [Boolean] True if the SAML Response is valid, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate(soft = true, request_id = nil)
```

Comment:
I added some comments about validation in lib/onelogin/ruby-saml/logoutresponse.rb I think that most if not all of them apply here.

There is enough overlap in the conceptual ideas across validation in these files that I question the repetition. I think even if nothing is changed with respect to my comments(no biggie) you should *definetly* pull out the shared validation logic into another file. By moving over the shared logic you will be able to more appropriately test, and debug issues.